feat: Add delegation credential error webhook trigger

Ryukemeister opened the pull request:
## What does this PR do?

Fixes CAL-6636

Implements a new webhook trigger `DELEGATION_CREDENTIAL_ERROR` that notifies organizations when delegation credential errors occur in Google Calendar or Office 365 calendar integrations. This webhook is restricted to organization webhooks only (API v2) and provides error details, credential info, and user email when delegation credential setup or authorization fails.

**Key changes:**
- Added `DELEGATION_CREDENTIAL_ERROR` to `WebhookTriggerEvents` enum in Prisma schema with migration
- Implemented webhook emission in Google Calendar and Office 365 calendar error handlers using standard `await Promise.allSettled()` pattern
- Added validation to restrict this trigger to organization webhooks only
- Created comprehensive e2e tests for CRUD operations with the new trigger
- Added webhook payload types and DTOs
- Added translation for the new trigger in locale files
- Fixed type safety issues: replaced "as any" type casting with safe type-narrowing helper
- Tightened `isNoShowPayload` type guard to prevent false positives with new payload type

**Link to Devin run**: https://app.devin.ai/sessions/a40ee3a076f349eea696237145a34831  
**Requested by**: rajiv@cal.com (@Ryukemeister)  
**Additional feedback from**: morgan@cal.com, alishaz-polymath

## Webhook Payload Structure

The webhook sends the following payload when a delegation credential error occurs:

```json
{
  "triggerEvent": "DELEGATION_CREDENTIAL_ERROR",
  "createdAt": "2025-01-01T00:00:00.000Z",
  "error": {
    "type": "CalendarAppDelegationCredentialInvalidGrantError",
    "message": "User might not exist in Google Workspace"
  },
  "credential": {
    "id": 123,
    "type": "google_calendar",
    "appId": "google-calendar"
  },
  "user": {
    "id": 456,
    "email": "user@example.com"
  }
}
```

Note: The `name` field was intentionally omitted from the user object per feedback - email is sufficient for identification.

## Mandatory Tasks

- [x] I have self-reviewed the code
- [x] I have updated the developer docs in /docs if this PR makes changes that would require a documentation change. **N/A** - This adds a new webhook trigger but doesn't change existing API behavior requiring doc updates.
- [x] I confirm automated tests are in place that prove my fix is effective or that my feature works.

## How should this be tested?

### Prerequisites
- An organization with delegation credentials configured for Google Workspace or Office 365
- API v2 access to create organization webhooks

### Test Steps

1. **Create an organization webhook with the new trigger:**
   ```bash
   POST /v2/organizations/{orgId}/webhooks
   {
     "subscriberUrl": "https://webhook.site/your-unique-url",
     "triggers": ["DELEGATION_CREDENTIAL_ERROR"],
     "active": true
   }
   ```

2. **Trigger the webhook by causing a delegation credential error:**
   - Set up a delegation credential with an invalid Client ID (Google Workspace)
   - Or set up a delegation credential for a user that doesn't exist in Azure AD (Office 365)
   - Attempt to create/view/update a calendar event using that credential

3. **Verify the webhook was called:**
   - Check your webhook endpoint received a POST request
   - Verify the payload contains error details, credential info, and user email
   - Confirm the `triggerEvent` is `DELEGATION_CREDENTIAL_ERROR`

4. **Verify organization-only restriction:**
   - Try creating a user webhook or event-type webhook with `DELEGATION_CREDENTIAL_ERROR` trigger
   - Should receive a 400 Bad Request: "DELEGATION_CREDENTIAL_ERROR trigger is only available for organization webhooks"

### E2E Tests

Run the e2e tests for organization webhooks:
```bash
cd apps/api/v2
yarn test:e2e:local --testPathPattern=organizations-webhooks.e2e-spec.ts
```

## Human Review Checklist

### Critical Items
- [x] **Type-narrowing helper in CalendarAuth.ts**: Verify the `getGoogleAuthErrorCode` function properly narrows the error type without using "as any". Check that it safely extracts error codes from the Google JWT library's error response structure.
- [x] **isNoShowPayload change**: Confirm that adding `&& "bookingUid" in data` is correct and won't break existing no-show webhooks. Verify that all `BookingNoShowUpdatedPayload` actually include `bookingUid` by checking `handleMarkNoShow.ts`.
- [ ] **Webhook emission locations**: Verify webhook emission is correctly implemented in all 5 delegation credential error scenarios:
  - Google Calendar: unauthorized_client, invalid_grant, and catch-all errors (CalendarAuth.ts:144-159)
  - Office 365: missing tenantId (CalendarService.ts:130-148), missing client credentials (CalendarService.ts:164-182, 205-223), user doesn't exist in Azure AD (CalendarService.ts:258-276)
- [ ] **Await and throw pattern**: Verify each of the 5 call sites properly (1) awaits the webhook emission, then (2) throws the original delegation error. The pattern should be: `await triggerDelegationCredentialErrorWebhook({...}); throw delegationError;`
- [ ] **Organization-only restriction**: Verify that `UserWebhooksService` and `EventTypeWebhooksService` correctly reject the `DELEGATION_CREDENTIAL_ERROR` trigger with appropriate error messages
- [ ] **Prisma migration**: Verify the migration file `20251110081648_add_delegation_credential_error_webhook_trigger/migration.sql` correctly adds the enum value

### Implementation Details
- [x] **Payload consistency**: Confirm the `name` field is removed from all type definitions: `DelegationCredentialErrorDTO`, `DelegationCredentialErrorPayloadType`, and the actual payload in `triggerDelegationCredentialErrorWebhook.ts`
- [x] **Error handling flow**: Verify that webhook emission errors are caught and logged but don't block the original delegation credential error from being thrown
- [x] **Repository pattern**: Confirm proper use of `WebhookRepository` instead of direct Prisma imports
- [x] **Async/await consistency**: Verify `getAuthUrl` method was correctly made async in Office365CalendarService and all 3 call sites (lines 86, 200, 290) properly await it

### Testing & CI
- [ ] **E2E test coverage**: Verify the 5 new e2e tests adequately cover create, retrieve, update, delete, and combined trigger scenarios
- [ ] **Test patterns**: Verify the tests follow the same patterns as existing tests in the file
- [ ] **Prisma types**: Confirm `yarn prisma generate` was run after schema changes

## Checklist

- [x] I have read the [contributing guide](https://github.com/calcom/cal.com/blob/main/CONTRIBUTING.md)
- [x] My code follows the style guidelines of this project
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have checked if my changes generate no new warnings

devin-ai-integration[bot] committed:
feat: add delegation credential error webhook trigger

- Add DELEGATION_CREDENTIAL_ERROR to WebhookTriggerEvents enum
- Create DelegationCredentialErrorDTO type for webhook payload
- Implement DelegationCredentialErrorWebhookService
- Add translation for delegation_credential_error
- Enable webhook for API v2 organization webhooks

This webhook will send delegation credential error data to configured URLs when errors occur during calendar authentication with delegation credentials.

Co-Authored-By: rajiv@cal.com <sahalrajiv6900@gmail.com>

devin-ai-integration[bot] committed:
fix: add delegation credential error payload type and type guards

- Add DelegationCredentialErrorPayloadType to sendPayload.ts
- Update WebhookPayloadType union to include new payload type
- Add isDelegationCredentialErrorPayload type guard function
- Update isEventPayload to exclude delegation credential errors
- Update template application logic to handle new payload type
- Add corresponding payload type to dto/types.ts for consistency

Co-Authored-By: rajiv@cal.com <sahalrajiv6900@gmail.com>

devin-ai-integration[bot] committed:
feat: add delegation credential error handling to WebhookNotificationHandler

- Add DELEGATION_CREDENTIAL_ERROR case to createPayload switch
- Return payload with error, credential, and user data
- Ensures exhaustive type checking passes for new trigger

Co-Authored-By: rajiv@cal.com <sahalrajiv6900@gmail.com>

devin-ai-integration[bot] committed:
feat: restrict DELEGATION_CREDENTIAL_ERROR to organization webhooks only

- Add validation in UserWebhooksService to reject DELEGATION_CREDENTIAL_ERROR
- Add validation in EventTypeWebhooksService to reject DELEGATION_CREDENTIAL_ERROR
- Ensures trigger is only available for API v2 organization webhooks as requested

Co-Authored-By: rajiv@cal.com <sahalrajiv6900@gmail.com>

devin-ai-integration[bot] committed:
feat: wire up delegation credential error webhook emission in calendar services

- Add webhook emission calls in CalendarAuth.ts for Google Calendar delegation errors
- Add webhook emission calls in Office365 CalendarService.ts for Azure AD delegation errors
- Implement actual webhook emission using WebhookRepository pattern
- Fix pre-existing lint warnings in Office365 CalendarService.ts (unused catch variables, unsafe optional chaining)

Co-Authored-By: rajiv@cal.com <sahalrajiv6900@gmail.com>

devin-ai-integration[bot] committed:
test: add e2e tests for DELEGATION_CREDENTIAL_ERROR webhook trigger

- Add comprehensive e2e tests for creating, retrieving, updating, and deleting webhooks with DELEGATION_CREDENTIAL_ERROR trigger
- Test combining DELEGATION_CREDENTIAL_ERROR with other triggers
- Fix import in triggerDelegationCredentialErrorWebhook.ts to use default import for sendPayload
- Tests follow existing patterns in organizations-webhooks.e2e-spec.ts

Co-Authored-By: rajiv@cal.com <sahalrajiv6900@gmail.com>

devin-ai-integration[bot] committed:
refactor: remove name field from webhook payload and delete unused service

- Remove name field from triggerDelegationCredentialErrorWebhook function signature and payload
- Update all call sites in GoogleCalendar and Office365 calendar services
- Update DelegationCredentialErrorDTO and DelegationCredentialErrorPayloadType to remove name field
- Delete unused DelegationCredentialErrorWebhookService.ts (dead code - not used anywhere)
- The helper function approach is more appropriate for app-store integrations without DI

Co-Authored-By: rajiv@cal.com <sahalrajiv6900@gmail.com>

devin-ai-integration[bot] committed:
refactor: make fire-and-forget intent explicit for webhook emission

- Add void cast to all triggerDelegationCredentialErrorWebhook calls
- Remove redundant .catch() handlers (helper already handles errors internally)
- This makes it clear that webhook emission is non-blocking by design
- Avoids delaying error propagation while webhook HTTP requests complete

Co-Authored-By: rajiv@cal.com <sahalrajiv6900@gmail.com>

devin-ai-integration[bot] committed:
feat: await webhook emission and add HTTP timeout for guaranteed delivery

- Change all triggerDelegationCredentialErrorWebhook calls from void to await
- Add 10-second timeout to webhook HTTP requests using AbortController
- Remove name field from DelegationCredentialErrorPayloadType to match payload
- Ensures webhooks are sent before error is thrown (per user requirement)
- Prevents indefinite hangs on unresponsive webhook endpoints

Co-Authored-By: rajiv@cal.com <sahalrajiv6900@gmail.com>

devin-ai-integration[bot] committed:
fix: make getAuthUrl async and await all call sites

- Make getAuthUrl async to support awaiting webhook emission
- Add await to all 3 getAuthUrl call sites (constructor, getAzureUserId, testDelegationCredentialSetup)
- Remove leftover name field from getAzureUserId webhook call
- Fixes TS1308 error about await in non-async function

Co-Authored-By: rajiv@cal.com <sahalrajiv6900@gmail.com>

devin-ai-integration[bot] committed:
docs: add JSDoc clarifying error handling guarantees for webhook trigger

Co-Authored-By: rajiv@cal.com <sahalrajiv6900@gmail.com>

devin-ai-integration[bot] committed:
Revert "docs: add JSDoc clarifying error handling guarantees for webhook trigger"

This reverts commit 3e33090197bfe2f2e3fb890402b32e04c44305e3.

devin-ai-integration[bot] committed:
Revert "feat: await webhook emission and add HTTP timeout for guaranteed delivery"

This reverts commit 9da7241f83a8373b4fadc03ccf34e097c28adf3a.

devin-ai-integration[bot] committed:
Revert "fix: make getAuthUrl async and await all call sites"

This reverts commit de28b7337149104412c861fd9b05e76fffc1fed7.

devin-ai-integration[bot] committed:
Revert "refactor: make fire-and-forget intent explicit for webhook emission"

This reverts commit f4f7fa06b7dfa151bfbea29905b8783261d9f353.

Ryukemeister committed:
Merge branch 'main' into devin/delegation-credential-errors-webhook-1762171203

devin-ai-integration[bot] committed:
feat: await webhook emission to match standard pattern

- Updated all webhook call sites to await triggerDelegationCredentialErrorWebhook
- Made getAuthUrl async and updated all 3 call sites to await it
- Removed .catch() wrappers at call sites (error handling is in trigger function)
- Matches standard pattern used in WebhookService.sendPayload with Promise.allSettled
- Ensures webhooks are sent before delegation errors are thrown

Co-Authored-By: rajiv@cal.com <sahalrajiv6900@gmail.com>

devin-ai-integration[bot] committed:
Merge branch 'devin/delegation-credential-errors-webhook-1762171203' of https://git-manager.devin.ai/proxy/github.com/calcom/cal.com into devin/delegation-credential-errors-webhook-1762171203

alishaz-polymath commented on packages/app-store/googlecalendar/lib/CalendarAuth.ts:
```diff
@@ -124,20 +125,38 @@ export class CalendarAuth {
-
-      if ((error as any).response?.data?.error === "invalid_grant") {
-        throw new CalendarAppDelegationCredentialInvalidGrantError(
+      } else if ((error as any).response?.data?.error === "invalid_grant") {
```
We can't have "as any", curious if you committed with `--no-verify` or not.

alishaz-polymath commented on packages/features/webhooks/lib/sendPayload.ts:
```diff
@@ -191,11 +212,19 @@ export function isOOOEntryPayload(data: WebhookPayloadType): data is OOOEntryPay
 
 export function isNoShowPayload(data: WebhookPayloadType): data is BookingNoShowUpdatedPayload {
-  return "message" in data;
+  return "message" in data && "bookingUid" in data;
```
Why did we make this change?

ThyMinimalDev committed:
Merge branch 'main' into devin/delegation-credential-errors-webhook-1762171203

devin-ai-integration[bot] committed:
fix: address PR feedback - add migration, remove 'as any', remove user.name

- Add Prisma migration for DELEGATION_CREDENTIAL_ERROR enum
- Replace 'as any' type casting with safe type-narrowing helper in CalendarAuth.ts
- Remove user.name field from DelegationCredentialErrorPayloadType (email is sufficient)
- Ensure all type definitions are consistent across sendPayload.ts and dto/types.ts

Addresses feedback from alishaz-polymath and morgan@cal.com

Co-Authored-By: rajiv@cal.com <sahalrajiv6900@gmail.com>

devin-ai-integration[bot] committed:
Merge branch 'devin/delegation-credential-errors-webhook-1762171203' of https://git-manager.devin.ai/proxy/github.com/calcom/cal.com into devin/delegation-credential-errors-webhook-1762171203

ThyMinimalDev committed:
chore: cleanup type in CalendarAuth

ThyMinimalDev committed:
Merge branch 'main' into devin/delegation-credential-errors-webhook-1762171203

ThyMinimalDev committed:
fix: missing DELEGATION_CREDENTIAL_ERROR in WEBHOOK_TRIGGER_EVENTS_GROUPED_BY_APP constant

ThyMinimalDev committed:
Merge branch 'main' into devin/delegation-credential-errors-webhook-1762171203

supalarry commented on packages/app-store/office365calendar/lib/CalendarService.ts:
```diff
@@ -123,12 +124,30 @@ export default class Office365CalendarService implements Calendar {
+              appId: this.credential.appId,
+            },
+            user: {
+              id: this.credential.userId ?? 0,
```
Here we fallback to 0, but surely if we have user with an email then we have a user with an id, so why not have check above
```
if(this.credential.user)
```

and then webhook payload user
```
 user: {
     id: this.credential.user.id,
     email: this.credential.user.email,
```

since email and id are defined on the User model.

ThyMinimalDev replied:
improved it but not great still, we don't have credential.user.id in calendar services

supalarry commented on packages/features/webhooks/lib/triggerDelegationCredentialErrorWebhook.ts:
```diff
@@ -0,0 +1,89 @@
+      return;
+    }
+
+    const payload = {
```
We can strongly type it `const payload: DelegationCredentialErrorPayloadType = {`

supalarry reviewed (changes requested):
Review:

ThyMinimalDev committed:
fix: review

ThyMinimalDev committed:
Merge branch 'main' into devin/delegation-credential-errors-webhook-1762171203

ThyMinimalDev committed:
fit: import webhook dto

ThyMinimalDev committed:
fit: type error

devin-ai-integration[bot] committed:
feat: add delegation credential error webhook handling to Office365 video adapter

- Emit webhook before throwing delegation credential errors in Office365 video
- Added webhook emission in 4 locations:
  1. Missing clientId/Secret in fetchNewTokenObject
  2. Missing tenantId in getAuthUrl
  3. Missing clientId/Secret in getAzureUserId
  4. User doesn't exist in Azure AD
- Made getAuthUrl async to support webhook emission
- Follows same pattern as GoogleCalendar and Office365Calendar implementations

Co-Authored-By: morgan@cal.com <morgan@cal.com>
Co-Authored-By: rajiv@cal.com <sahalrajiv6900@gmail.com>

devin-ai-integration[bot] committed:
Merge branch 'devin/delegation-credential-errors-webhook-1762171203' of https://git-manager.devin.ai/proxy/github.com/calcom/cal.com into devin/delegation-credential-errors-webhook-1762171203

ThyMinimalDev committed:
fixup! Merge branch 'devin/delegation-credential-errors-webhook-1762171203' of https://git-manager.devin.ai/proxy/github.com/calcom/cal.com into devin/delegation-credential-errors-webhook-1762171203

ThyMinimalDev committed:
Merge branch 'main' into devin/delegation-credential-errors-webhook-1762171203
